pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +214,195 @@ pub mod pallet {
 			todo!()
 		}
 
-		// TODO: not sure if we really need that. Check that out with Raphael.
 		#[pallet::call_index(6)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, Some(who))
 		}
 
 		#[pallet::call_index(7)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn force_start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn force_start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, None)
 		}
 
-		// todo: check if we really need that tx.
 		#[pallet::call_index(8)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn destroy_accounts(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn refund_account(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			account: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let who = T::Lookup::lookup(account)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			let pool_account = pool_id.clone().into();
+
+			let mut total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::ZeroAmount);
+
+			let total_issuances: Vec<(FungiblesAssetIdOf<T>, FungiblesBalanceOf<T>)> = pool_details
+				.bonded_currencies
```

Comment:
This is gonna be hard to benchmark accurately. I think it would be better to also specify the bonded currency ID in the extrinsic. In that case both testing and benchmarking becomes almost trivial.

Reply:
The more atomic we make the refund extrinsic though, the more often we have to look up the supply of all tokens and of the collateral. In the current implementation, we do it once per account, if changed as you suggested, we'd have to do it once per account per currency.

Reply:
If you think this solution is not viable due to benchmarking restrictions, that's what we'll have to do though.

Reply:
All storage reads and writes go through an intermediate overlay layer, which means that calling the same extrinsic with a different currency ID (if done within the same block, with e.g. a batch), especially since all currencies are stored within the same storage entry inside a vector, won't have any worse performance than what you are doing here. It's just much, much simpler to think about and to benchmark. Also, you can then directly check whether the currency has zero balance and do something with it, instead of having a somewhat complicated loop and a flag as you did above, if I understood correctly what the flag in the loop is supposed to do.

Reply:
This ended up being a larger refactor; see d737d8a5237ad85eadd0e8787493334e0dbb89d7

I now introduced a dedicated refunding state, so that the flow is now:
- The pool owner calls `start_refund`. This would fail if the collateral or total supply of bonded tokens is 0 - in this case the owner should skip right to `start_destroy`.
- Once in refunding state, anyone can call `refund_account` for any account.
- When the collateral of total supply drop to 0, the pool owner can call `start_destroy`. This fails otherwise.
- Once in destroying state, anyone can call `destroy_accounts` and `destroy_approvals` on the assets pallet (if there are any left)
- If there are no more accounts or approvals, anyone can call `finish_destroy`.

Reply:
This looks like a more runtime-friendly solution. I will review this in my second review, which is coming. I'll mark this as resolved and in case I find something else, I'll make a new comment.